Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp

```diff
@@ -0,0 +1,173 @@
+{
+    return adoptRef(*new FileSystemWritableFileStreamSink(source));
+}
+
+FileSystemWritableFileStreamSink::FileSystemWritableFileStreamSink(FileSystemFileHandle& source)
+    : m_source(source)
+{
+}
+
+FileSystemWritableFileStreamSink::~FileSystemWritableFileStreamSink()
+{
+    if (!m_isClosed)
+        protectedSource()->closeWritable(false);
+}
+
+// https://fs.spec.whatwg.org/#write-a-chunk
+void FileSystemWritableFileStreamSink::write(ScriptExecutionContext& context, JSC::JSValue value, DOMPromiseDeferred<void>&& promise)
+{
+    if (m_isClosed)
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
+
+    if (value.isUndefined())
+        return promise.reject(Exception { ExceptionCode::TypeError, "Data is undefined"_s });
+
+    auto scope = DECLARE_THROW_SCOPE(context.vm());
+    auto chunkResult = convert<IDLUnion<IDLArrayBufferView, IDLArrayBuffer, IDLInterface<Blob>, IDLUSVString, IDLDictionary<FileSystemWritableFileStream::WriteParams>>>(*context.globalObject(), value);
+    if (UNLIKELY(chunkResult.hasException(scope)))
+        return promise.reject(Exception { ExceptionCode::ExistingExceptionError });
+
+    auto writeParams = writeParamsFromChunk(chunkResult.releaseReturnValue());
+    switch (writeParams.type) {
+    case FileSystemWriteCommandType::Seek: {
+        if (!writeParams.position)
+            return promise.reject(Exception { ExceptionCode::TypeError, "Position is not provided for seek"_s });
+
+        protectedSource()->executeCommandForWritable(writeParams.type, writeParams.position, writeParams.size, { }, WTFMove(promise));
+        return;
+    }
+    case FileSystemWriteCommandType::Truncate: {
+        if (!writeParams.size)
+            return promise.reject(Exception { ExceptionCode::TypeError, "Size is not provided for truncate"_s });
+
+        protectedSource()->executeCommandForWritable(writeParams.type, writeParams.position, writeParams.size, { }, WTFMove(promise));
+        return;
+    }
+    case FileSystemWriteCommandType::Write: {
+        if (!writeParams.data)
+            return promise.reject(Exception { ExceptionCode::TypeError, "Data is not provided for write"_s });
+
+        fetchDataBytesForWrite(*writeParams.data, [this, protectedThis = Ref { *this }, promise = WTFMove(promise), type = writeParams.type, size = writeParams.size, position = writeParams.position](bool fetched, auto dataBytes) mutable {
+            if (m_isClosed)
+                return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
+
+            if (!fetched)
+                return promise.reject(Exception { ExceptionCode::TypeError, "Failed to read data"_s });
+
+            protectedSource()->executeCommandForWritable(type, position, size, dataBytes, WTFMove(promise));
+        });
+    }
+    }
+}
+
+void FileSystemWritableFileStreamSink::close()
+{
+    if (m_isClosed)
+        return;
```

Comment:
Do we have a test where we write then close?
It seems to me that normally, we would finish writing and then closing.
Which does not seem to be the case if we write a blob say.

Reply:
> Do we have a test where we write then close?

Yes, the new test does write and close.